### PR TITLE
Issue/4889 notifications read mark

### LINKED ISSUE #4889: Notifications: Opening WP notif from Android system notification doesn't appear as Read
### Steps to reproduce the behavior
1) Spawn a WP.com notification, such as by leaving a comment on your site as another user
2) Open the Android notification, which should launch the app to the WP notification
3) Back out to the notifications list

### Expected behavior
To have the notification marked as read

### Actual behavior
The notification still appears unread.  It's a display issue only, looking at the notification on another device it appears read (or if you refresh the in-app notifications list)

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4889 by re-marking the notification as read once the NotificationUpdateService has finished doing its thing, as per [this comment](https://github.com/wordpress-mobile/WordPress-Android/issues/4889#issuecomment-267701075)

To test:

0. Send the Android app to the background (i.e. press menu button on your device)
1. Spawn a WP.com notification, such as by leaving a comment on your site as another user
2. Open the Android notification, which should launch the app to the WP notification
3. Back out to the notifications list

The note should now be marked read.

cc @daniloercoli

### CODE REVIEW

**Comment by daniloercoli** [comment]
Wondering if we can just delay the execution of `NotificationUpdateService`, instead on setting up this hack around a race condition issue?

**Comment by mzorz** [comment]
> Wondering if we can just delay the execution of NotificationUpdateService, instead on setting up this hack around a race condition issue?
(While we're waiting for a more robust solution that addresses all of this kind of problems).

It think there are basically 3 approaches to tackling race conditions, generally speaking: 
1. either trying to delay or prevent one thread from happening,
2. adding flagging that attempts to control the execution flow
3. re-thinking architecture, which is what I believe you're referring to by a more robust solution

For point 1, I think delaying the execution of something while hoping for another process to finish without coordination is risky and can't assure this can be repeated 100% of times.
For point 2, the cost of adding such coordination is high and would add too much complexity for what it's worth.
For point 3, there are various levels. I focused on what other things are already happening in code, that can serve to me as an indicative of something that is happening (like, is this thread finished? when will it be finished? what will the state of things be in the end? etc)

I think this PR focuses on using things that we know will happen, like:
1. we know the update service is being triggered by a specific intent. If the intent carries on a `noteId`extra, then we can be 100% sure this has been because the user tapped on a Notification, and not any other cause (*)
2. If that was the intent, then we can rest assured the user attempted to read the body for such notification, and then application acted in consequence (actually showed the notification to the user). That means, we can be 100% sure the notification was read, which is the main event that we need to keep track of for the read/unread feature.
3. Finally, if we know the update service has been brought to life by tapping on a notification, then such Note will likely be marked `unread` in the data coming from the server, but even regardless of what value it carries, the fact that the user is reading the notification gives us enough reason to mark it `read` when the update service call is finished.

I’m happy with this solution as it’s deterministic (as opposed to hoping something will happen, or trying to control execution flow for something that is inherently indeterministic like connection response times are).

(*) Maybe we could make things more explicit, instead of just taking the `noteId` in the intent to piggy back on it we could add another extra parameter to the intent bundle like “IS_TAPPED_ON_NOTIFICATION” boolean and send it through to the service (along with the noteid) to aim for code readability.

Let me know your thoughts? 🙇

**Comment by daniloercoli** [comment]
>3.  re-thinking architecture, which is what I believe you're referring to by a more robust solution

I was thinking to add a new kind of push notifications type, say "notes_update", that has the IDs of the notes, and the action we should apply on them. The server send this new kind of PN to the clients, that react by updating the notes in their DB. This will fix this issues once and for all.

If the user has push notifications disabled on the device this issue is not there.

File: WordPress/src/main/java/org/wordpress/android/WordPress.java [hunk 1]
```diff
@@ -846,7 +848,12 @@ private void onAppComesFromBackground() {
                 // Refresh account informations and Notifications
 
                 if (AccountHelper.isSignedInWordPressDotCom()) {
-                    NotificationsUpdateService.startService(getContext());
+                    if (activity.getIntent().hasExtra(NotificationsListFragment.NOTE_ID_EXTRA)) {
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/WordPress.java]: Is it possible to have a null `Intent` here?

**Comment by daniloercoli** [comment]
> (*) Maybe we could make things more explicit, instead of just taking the noteId in the intent to piggy back on it we could add another extra parameter to the intent bundle like “IS_TAPPED_ON_NOTIFICATION” boolean and send it through to the service (along with the noteid) to aim for code readability.

Yes, this is probably better for future reading of the code.

**Comment by mzorz** [comment]
> I was thinking to add a new kind of push notifications type, say "notes_update", that has the IDs of the notes, and the action we should apply on them. The server send this new kind of PN to the clients, that react by updating the notes in their DB. This will fix this issues once and for all.

If I'm following correctly (I'm not sure, I'm asking) the proposed flow would go like this: whenever an user reads a notification, the client app (web or mobile) calls the markAsRead API endpoint, and internally that endpoint not only marks the note read in the db, but also triggers a `notes_update` type of notification for all registered devices for that user, correct?
But, when the user reads the note on the device, while we need to tell the server, there's no need to wait for the action to come back in that new `notes_update` PN internally. We know we need to mark it read right away in our local db to reflect changes, right?

Are you thinking of having that and replacing the current NotificationsUpdateService? 

1. If it's not a replacement: *Still, we could have a race condition*
Also, if in between getting the `notes_update` PN you open the app thus triggering the NotificationsUpdateService and such call ends _after_ receiving the PN in point 2, we would end up having the same problem. (assuming we are not ditching the service)

2. If it is a replacement: *What if a device misses an update?*
Given the proposed solution relies on PNs being sent/received, this could lead to potential inconsistency as some clients might get the read signal, while others might not get it. On the contrary, pulling every now and then as we're doing right now seems to take care of this more reliably (as we actually clear the notifications db and re-insert whatever comes from the server, making a clean new copy each time).

I don't know maybe you were thinking of the `notes_update` PN as an additional thing to the current pulling? but having the 2 mechanisms together would bring us to point 2 above

File: WordPress/src/main/java/org/wordpress/android/WordPress.java (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/WordPress.java]: While theoretically possible, I think it's quite tricky for it to happen here, anyway good catch addressed in fee0f38

**Comment by mzorz** [comment]
>> (*) Maybe we could make things more explicit, instead of just taking the noteId in the intent to piggy back on it we could add another extra parameter to the intent bundle like “IS_TAPPED_ON_NOTIFICATION” boolean and send it through to the service (along with the noteid) to aim for code readability.

> Yes, this is probably better for future reading of the code.

addressed in 94277d7

**Comment by daniloercoli** [comment]
> If I'm following correctly (I'm not sure, I'm asking) the proposed flow would go like this: whenever an user reads a notification, the client app (web or mobile) calls the markAsRead API endpoint, and internally that endpoint not only marks the note read in the db, but also triggers a notes_update type of notification for all registered devices for that user, correct?

I wasn't thinking to radical changes on the client. Just a lightweight solution that runs on the backend, and notifies clients when a note is marked as read (or updated, or deleted). 
The Android client, like it does at the moment, marks the note read in the db, and pings the remote server, that in order sends a PN to all clients.
More or less what's already happening with the badge reset note, and more or less what Pinghub will do. 

>If it's not a replacement: Still, we could have a race condition
if in between getting the notes_update PN you open the app thus triggering the NotificationsUpdateService and such call ends after receiving the PN in point 2, we would end up having the same problem. (assuming we are not ditching the service)

Yes, race conditions could still happen, but it's unlikely that:
- The client send the *markAsRead* call
- The client send the *getNotes* call
- The server receive first the *getNotes* call and replies to the clients
- The server mark the note as read and send the push notifications
- Push notifications are lotst.

**Comment by daniloercoli** [comment]
Will merge this PR tomorrow.

**Comment by mzorz** [comment]
> More or less what's already happening with the badge reset note, and more or less what Pinghub will do.

Nice! I almost forgot we were already doing this with the badge reset note 😁 

Yeah the problem this PR tries to address is a consequence of having to pull data. If we pull data and write locally, things can happen (things not synced). If we move to pinghub things are better, but still  these all aren't big problems anyway 😀

**Pull request merged**